File: comparisons/lmer_weighted.py
"""
Weighted LMM parity script using statsmodels.

`statsmodels.regression.mixed_linear_model.MixedLM.fit` does not apply prior
case weights the way `lme4::lmer(..., weights=)` does. To keep this script
self-contained (no `lme_python` wheel), we approximate prior weights by
replicating each row a number of times proportional to its weight, then fit
REML MixedLM on the expanded table. Coefficients will not match R/Rust exactly;
use those implementations for numerical parity.
"""

from __future__ import annotations

import os
import sys

import numpy as np

try:
    import pandas as pd
    import statsmodels.formula.api as smf
except ImportError:
    print("Please install pandas and statsmodels to run this example:")
    print("pip install pandas statsmodels numpy")
    sys.exit(1)


def main() -> None:
    path = os.path.join("tests", "data", "sleepstudy.csv")
    if not os.path.exists(path):
        print(f"Could not find the dataset at {path}")
        print("Please run this example from the root of the lme-rs repository.")
        sys.exit(1)

    data = pd.read_csv(path)
    n = len(data)
    w = np.array([0.5 + (i % 5) * 0.1 for i in range(n)], dtype=float)
    mult = np.maximum(1, (w / w.min() * 8.0).round().astype(int))
    expanded = data.loc[np.repeat(data.index.values, mult)].reset_index(drop=True)

    print("Loading tests/data/sleepstudy.csv ...")
    print(f"Expanded row count for approximate weights: {len(expanded)} (from {n})")
    print("\nFitting model: Reaction ~ Days + (Days | Subject) [REML, expanded-data proxy]")

    model = smf.mixedlm(
        "Reaction ~ Days",
        expanded,
        groups=expanded["Subject"],
        re_formula="~Days",
    )
    result = model.fit(reml=True)
    print("\n=== Model Summary ===")
    print(result.summary())

    print("\n=== Predictions (population-level, fixed part only) ===")
    fe = result.fe_params
    days = np.array([0.0, 1.0, 5.0, 10.0])
    preds = fe["Intercept"] + fe["Days"] * days
    print(preds)


if __name__ == "__main__":
    main()

File: comparisons/lmer_weighted.rs
use lme_rs::lmer_weighted;
use ndarray::Array1;
use polars::prelude::*;
use std::path::PathBuf;

fn main() -> anyhow::Result<()> {
    let file_path = PathBuf::from("tests").join("data").join("sleepstudy.csv");

    if !file_path.exists() {
        eprintln!("Could not find the dataset at {}", file_path.display());
        eprintln!("Please run this example from the root of the lme-rs repository.");
        std::process::exit(1);
    }

    println!("Loading data from {}...", file_path.display());

    let mut file = std::fs::File::open(&file_path)?;
    let df = CsvReadOptions::default()
        .with_has_header(true)
        .into_reader_with_file_handle(&mut file)
        .finish()?;

    println!("Successfully loaded {} rows.", df.height());

    let formula = "Reaction ~ Days + (Days | Subject)";
    println!("\nFitting weighted model: {}", formula);
    println!(
        "Prior weights w_i = 0.5 + (row_index mod 5) * 0.1 (same pattern as `benches/bench_math`)"
    );
    println!("Evaluating Restricted Maximum Likelihood (REML)...");

    let n = df.height();
    let weights = Array1::from_vec((0..n).map(|i| 0.5_f64 + (i % 5) as f64 * 0.1).collect());

    let fit = lmer_weighted(formula, &df, true, Some(weights))?;

    println!("\n=== Model Summary ===");
    println!("{}", fit);

    println!("\n=== Predictions ===");
    println!("Generating predictions for Subject 308 at Days 0, 1, 5, and 10...");

    let new_days = Series::new("Days".into(), &[0.0, 1.0, 5.0, 10.0]);
    let new_subject = Series::new("Subject".into(), &["308", "308", "308", "308"]);
    let newdata: DataFrame = DataFrame::new(vec![new_days.into(), new_subject.into()])?;

    let preds = fit.predict(&newdata)?;
    println!("Predictions:\n{:?}", preds);

    Ok(())
}
